[Tan Zi Yang] iP

ziyang27 opened the pull request:
# Dibo the Task Chatbot 

Dibo frees your mind from having to remember things you need to do. It's
- test-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

> [!IMPORTANT]
> Take a look at the list of commands and their behaviors to familiarise yourself for **MAXIMUM SPEED!**

All you need to do is,
1. download it from [here](https://github.com/ziyang27/ip).
2. double-click it.
3. add your tasks.
4. sit back and let it manage your task for you :100:

Most importantly, it is totally **FREE!**

Key features:
- [x] Managing tasks
- [x] Saving tasks to harddrive
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

***
Before we begin, let's look at some of the ***COOL*** 😎 commands!
- `list` lists all the tasks in the task list, their completion status, type, and other relevant information.
- `mark` and `unmark` to mark the respective task as done or not.
- `delete` deletes the task from the task list.

***
In case you are wondering, here's the main method:
```java
public static void main(String[] args) {
        Dibo dibo = new Dibo("./data/dibo.txt");
        dibo.run();
}
```
**What? Just like that?** 
Yes, click in to find out more! :monocle_face:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ziyang27 committed:
Rename, Greet, and Exit

ziyang27 committed:
Echo

ziyang27 committed:
Add, List

ziyang27 committed:
Mark as Done

ziyang27 committed:
ToDos, Events, Deadlines

ziyang27 committed:
Automated Text UI Testing

ziyang27 committed:
Handle Errors

ziyang27 committed:
Delete

ziyang27 committed:
Update program logic

ziyang27 committed:
Add saving feature

ziyang27 committed:
Add saving feature

ziyang27 committed:
Add date format

ziyang27 committed:
Edit dibo.text

ziyang27 committed:
Merge branch 'branch-level-8'

ziyang27 committed:
Add more OOP

ziyang27 committed:
Create dibo package

ziyang27 committed:
Add dibo.text

ziyang27 committed:
Merge branch 'add-gradle-support'

ziyang27 committed:
Edit OOP and packages

ziyang27 committed:
Set up gradle

ziyang27 committed:
Check style

ziyang27 committed:
Add JUnit tests

ziyang27 committed:
Add JavaDoc

ziyang27 committed:
Ensure coding standard

ziyang27 committed:
Add find feature

ziyang27 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/dibo/Ui.java

ziyang27 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/dibo/Dibo.java
#	src/main/java/dibo/task/Task.java

ziyang27 committed:
Edit import style

ziyang27 committed:
Add additional checks for tasks

ziyang27 committed:
Use CheckStyle

SherisseTJW commented on src/main/java/dibo/Dibo.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public static void main(String[] args) {
+        Dibo dibo = new Dibo("./data/dibo.txt");
```
It might be better to store the filepath as a private final field

SherisseTJW commented on src/main/java/dibo/Parser.java:
```diff
@@ -0,0 +1,146 @@
+     */
+    public static Command parse(String fullCommand) throws DiboException {
+        String[] commandDetails = fullCommand.split(" ");
+        switch (commandDetails[0]) {
```
Maybe you can add the `break;` statement for all the cases or have a `// Fallthrough` comment?

SherisseTJW commented on src/main/java/dibo/Parser.java:
```diff
@@ -0,0 +1,146 @@
+        }
+    }
+
+    private static boolean hasDescription(String fullCommand) {
+        String[] details = fullCommand.split("/");
+        String[] typeAndDescription = details[0].split(" ");
+        return typeAndDescription.length == 2;
+    }
+
+    private static String getDescription(String fullCommand, String type) {
+        String[] details = fullCommand.split("/");
+        return details[0].substring(type.length()).trim();
+    }
+
+    private static LocalDate getByDate(String fullCommand)
+            throws DateTimeParseException, IndexOutOfBoundsException {
+        String[] details = fullCommand.split("/");
+        String byString = details[1].substring(2).trim();
+        return LocalDate.parse(byString, Parser.INPUT_FORMAT);
+    }
+    private static LocalDate getStartDate(String fullCommand)
+            throws DateTimeParseException, IndexOutOfBoundsException {
+        String[] details = fullCommand.split("/");
+        String startString = details[1].substring(4).trim();
+        return LocalDate.parse(startString, Parser.INPUT_FORMAT);
+    }
+
+    private static LocalDate getEndDate(String fullCommand)
+            throws DateTimeParseException, IndexOutOfBoundsException {
+        String[] details = fullCommand.split("/");
+        String endString = details[2].substring(2).trim();
+        return LocalDate.parse(endString, Parser.INPUT_FORMAT);
+    }
```
Not sure I like how these helper methods are `private static` rather than just `private`. Is there a rationale behind this?

ziyang27 replied:
Actually, I didn't take notice of this before. I will take note in the future! Thank you so much 😄

SherisseTJW commented on src/main/java/dibo/Storage.java:
```diff
@@ -0,0 +1,123 @@
+                Task task;
+                switch (type) {
+                case "todo":
+                    String description1 = details[2];
+                    task = new ToDo(description1);
+                    break;
+                case "deadline":
+                    String description2 = details[2];
+                    String by = details[3].trim();
+                    task = new Deadline(description2, convertToLocalDate(by));
+                    break;
+                case "event":
+                    String description3 = details[2];
```
Not sure I like the name `description1`, perhaps you can rename it to `todoDescription` instead? Or have a singular `description` declaration at line 50 above which can be used for each of the cases since you have `description2` and `description3` as well?

SherisseTJW commented on src/main/java/dibo/TaskList.java:
```diff
@@ -0,0 +1,144 @@
+    public String getDisplayFormat() {
+        StringBuilder list = new StringBuilder("Here are the tasks in your list:\n");
+        for (int i = 0; i < count; ++i) {
+            list.append((i + 1));
+            list.append(".");
+            list.append(tasks.get(i).toString());
+            list.append("\n");
```
Perhaps you can consider using `String.format()` to have it in 1 line?

ziyang27 replied:
Thank you for the suggestion! I will look into it!

SherisseTJW reviewed (commented):
Overall, LGTM!

Just some minor suggestions to clean up the code ^ ^

ziyang27 committed:
Add GUI

ziyang27 committed:
Edit text file

ziyang27 committed:
Merge branch 'branch-Level-10'

ziyang27 committed:
Add varargs

ziyang27 committed:
Edit build file to create new jar file

The main class is no longer Dibo.java and has changed to Main.java inside the dibo package. The base name is still 'duke' instead of 'dibo'.

Let's,
* change the entry point
* update the base name

This enables the new jar file to function and provides more clarity for future users.

ziyang27 committed:
Add assertions to various points of the code

There are some assumptions within the code that are not tested for which may result in runtime error.

Lets,
* add an assertion to test the return output function in the Ui class
* add assertions to test that the TaskList class interact correctly with the Task class

Using assertions at these aforementioned points help to ensure that the code is robust and behaving as intended.

ziyang27 committed:
Merge pull request #1 from ziyang27/branch-A-Assertions

Add assertions to various points of the code

ziyang27 committed:
Remove magic literals in Ui class

ziyang27 committed:
Touch up code quality in TaskList class

ziyang27 committed:
Break up the loadData function in Storage class

ziyang27 committed:
Break up parse function in Parser class

ziyang27 committed:
Touch up for code quality

ziyang27 committed:
Merge pull request #2 from ziyang27/branch-A-CodeQuality

Branch a code quality

ziyang27 committed:
Change the hasKeywords function in Task class to use stream

ziyang27 committed:
Merge pull request #3 from ziyang27/branch-A-Streams

Change the hasKeywords function in Task class to use stream

ziyang27 committed:
Set up CI

ziyang27 committed:
Add a new task, do-after

ziyang27 committed:
Merge pull request #4 from ziyang27/branch-B-DoAfterTasks

Add a new task, do-after

JonChong98 reviewed (commented):
Can't find any glaring issues with code quality aside from the minor ones already highlighted by fellow reviewer. Good job!

ziyang27 committed:
Allow resizing of the Window

ziyang27 committed:
Add colors

ziyang27 committed:
Change images to fit personality

ziyang27 committed:
Add function to create new file when file does not exist

ziyang27 committed:
Add more tests for the two classes

ziyang27 committed:
Improve JavaDoc comments

ziyang27 committed:
Give credit for reusing course materials, for practice

ziyang27 committed:
Add user guide

ziyang27 committed:
Add user guide

ziyang27 committed:
Merge branch 'master' of https://github.com/ziyang27/ip

ziyang27 committed:
Edit user guide

ziyang27 committed:
Update README.md

ziyang27 committed:
Update README.md

ziyang27 committed:
Edit README

ziyang27 committed:
Merge branch 'master' of https://github.com/ziyang27/ip

ziyang27 committed:
Update README.md

ziyang27 committed:
Include alert bar for README

ziyang27 committed:
Merge branch 'master' of https://github.com/ziyang27/ip

ziyang27 committed:
Improve formatting

ziyang27 committed:
Change main class to Launcher

ziyang27 committed:
Edit README

ziyang27 committed:
Remove magic literals in Parser

ziyang27 committed:
Include tests for hasKeywords function in Event class
